File: apps/control-api/src/lib/health.ts
import { Elysia } from "elysia";

// ADR-026: every service exposes /health/live and /health/ready with
// the same meaning. Liveness never touches a dependency (a slow
// Postgres must never make a live process look dead to an
// orchestrator's liveness probe) — same shape as
// services/data-plane's internal/platform/health.Register, just in
// TypeScript for control-api.
export type Checker = () => Promise<void>;

interface CheckResult {
	status: "up" | "down";
	error?: string;
}

// healthRoutes returns a standalone Elysia plugin — mount it on the
// *unprefixed* root app, not inside the /api/v1-prefixed sub-app
// (confirmed empirically: a plugin merged via .use() inherits its
// parent's prefix in Elysia 1.4, so /health/live would otherwise
// become /api/v1/health/live, which no orchestrator or `docker
// healthcheck` config expects).
export function healthRoutes(checks: Record<string, Checker>) {
	return new Elysia()
		.get("/health/live", () => new Response(null, { status: 200 }))
		.get("/health/ready", async ({ set }) => {
			const results: Record<string, CheckResult> = {};
			let ready = true;

			await Promise.all(
				Object.entries(checks).map(async ([name, check]) => {
					try {
						await check();
						results[name] = { status: "up" };
					} catch (err) {
						ready = false;
						results[name] = {
							status: "down",
							error: err instanceof Error ? err.message : String(err),
						};
					}
				}),
			);

			set.status = ready ? 200 : 503;
			return { checks: results };
		});
}

File: services/data-plane/internal/platform/health/health.go
// Package health wires the two endpoints ADR-026 requires every
// service to expose: /health/live and /health/ready. Shared here
// rather than duplicated per cmd, the way internal/platform/config
// already splits Load per-binary but shares the loader mechanics.
package health

import (
	"context"
	"encoding/json"
	"net/http"
)

// Checker reports whether one dependency this service relies on is
// currently reachable. Each service supplies its own set — ingestion
// checks Postgres, the worker checks ClickHouse + Redpanda — readiness
// is never the same fixed check for every binary.
type Checker func(ctx context.Context) error

type checkResult struct {
	Status string `json:"status"`
	Error  string `json:"error,omitempty"`
}

type readyResponse struct {
	Checks map[string]checkResult `json:"checks"`
}

// Register wires /health/live and /health/ready onto mux.
//
// Liveness never touches a dependency — operations.md's graceful-
// shutdown section draws the same distinction other cloud-native
// services make: a slow Postgres must never make a live process look
// dead to an orchestrator's liveness probe (which would restart a
// perfectly healthy process for someone else's outage).
//
// Readiness runs every checker and reports which ones failed —
// 200 only if all of them succeed, 503 otherwise, with a per-
// dependency breakdown so "what's actually down" is answered directly
// rather than needing a follow-up investigation.
func Register(mux *http.ServeMux, checks map[string]Checker) {
	mux.HandleFunc("GET /health/live", func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusOK)
	})

	mux.HandleFunc("GET /health/ready", func(w http.ResponseWriter, r *http.Request) {
		resp := readyResponse{Checks: make(map[string]checkResult, len(checks))}
		ready := true

		for name, check := range checks {
			if err := check(r.Context()); err != nil {
				ready = false
				resp.Checks[name] = checkResult{Status: "down", Error: err.Error()}
				continue
			}
			resp.Checks[name] = checkResult{Status: "up"}
		}

		w.Header().Set("Content-Type", "application/json")
		if ready {
			w.WriteHeader(http.StatusOK)
		} else {
			w.WriteHeader(http.StatusServiceUnavailable)
		}
		_ = json.NewEncoder(w).Encode(resp)
	})
}
